exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java

```diff
@@ -435,9 +427,14 @@ public synchronized boolean reconnect() {
   }
 
   private void connect(DrillbitEndpoint endpoint) throws RpcException {
-    final FutureHandler f = new FutureHandler();
-    client.connect(f, endpoint, props, getUserCredentials());
-    f.checkedGet();
+    client.connect(endpoint, properties, getUserCredentials()).checkedGet();
+    if (client.serverRequiresAuthentication()) {
+      try {
+        client.authenticate(null).checkedGet();
+      } catch (SaslException e) {
+        throw new RpcException(e);
+      }
```

Comment:
Originally if client fail's to connect to server because of Authentication failure, then it doesn't retry since `InvalidConnectionInfoException` is thrown and handled separately.

But If authentication using SASL fails then we are always throwing `RpcException` and we will end up doing retry. We should have a check for exception thrown by SASL authentication failure and doesn't do retry in that scenario as well.